File: CrackingTheCodingInterview/1.1/isUnique.java
import java.util.HashMap;

public class OnePointOne {
    //Implement an algorithm to determine if a string has all unique characters.

    public static Boolean isUnique(String input){
        String[] inputArray = input.split("");
        HashMap<String, Boolean> charMap = new HashMap<>();
        int inputArrayLength = inputArray.length;
        for(int i=0; i<inputArrayLength; i++){
            if(charMap.get(inputArray[i]) == null){
                charMap.put(inputArray[i], true);
            }else{
                return false;
            }
        }
        return true;
    }

    //use main to test
    public static void main(String[] args){
        System.out.println(OnePointOne.isUnique("abc"));
        System.out.println(OnePointOne.isUnique("pancakes"));
    }
}

File: CrackingTheCodingInterview/1.1/isUnique.js
function isUnique(input){

    hashMap = {};
    inputLength = input.length;
    for(var i=0; i<inputLength; i++){
        if(hashMap[input[i]]==null){
            hashMap[input[i]] = true;
        }else{
            return false;
        }
    }
    return true;

}

// without map 
function isUniqueNoDataStructures(input){

    for(var i=0; i<input.length; i++){
        for(var j=i+1;j<input.length; j++){
            if(input[i] === input[j]){
                return false;
            }
        }
    }
    return true;

}

console.log('pancakes');
console.log(isUnique('pancakes'));
console.log('able');
console.log(isUnique('able'));
